File: java/nCr_nPr.java
static int[] facts;
static int mod = (int) 1e9+7;
static int mul(int a, int b)
{
    return (int) ((long) a * b % mod);
}
static int pow(int a, int b)
{
    if (b == 0) return 1;
    if (b % 2 == 1) return mul(a, pow(a, b - 1));
    int c = pow(a, b / 2);
    return mul(c, c);
}
static int inv(int x)
{
    return pow(x, mod-2);
}
static void calc_fac(int m)
{
    facts = new int[m + 1];
    facts[0] = 1;
    for (int i = 1; i <= m; i++) {
        facts[i] = mul(facts[i-1], i);
    }
}
static int nPermuteK(int n, int k)
{
    return mul(facts[n], inv(facts[n-k]));
}
static int nChooseK(int n, int k) 
{
    return mul(nPermuteK(n, k), inv(facts[k]));
}

File: python/nCr_nPr.py
mod = int(1e9 + 7)
nax = int(2e5 + 10)


def mul(a: int, b: int) -> int:
    return (a * b) % mod


def pow(a: int, b: int) -> int:
    if b == 0:
        return 1
    if b % 2 == 1:
        return mul(a, pow(a, b - 1))
    else:
        c = pow(a, b // 2)
        return mul(c, c)


def inv(x: int) -> int:
    return pow(x, mod - 2)


facts = [1] * (nax + 1)
for i in range(1, nax):
    facts[i] = mul(facts[i - 1], i)


def nPr(n: int, r: int) -> int:
    return mul(facts[n], inv(facts[n - r]))


def nCr(n: int, r: int) -> int:
    return mul(nPr(n, r), inv(facts[r]))